In [209]:
# get individual worm data
# Conny Lin | June 1, 2020

In [1]:
import numpy as np
import pandas as pd
import os
from pathlib import Path
from pathlib import PurePath
import pathlib
import glob
import pickle

In [2]:
dir_save = '/Users/connylin/Dropbox/CA/ED _20200119 Brain Station Data Science Diploma/Capstone/data'

# Search DB for plates

In [3]:
# load file summary
filename = os.path.join(dir_save, 'file_summary_mwt.pickle')
with open(filename,'rb') as f:
    file_summary = pickle.load(f)

In [4]:
file_summary.shape

(7298, 38)

In [214]:
file_summary

filepath                    \
suffix                                                      .blob .blobs .csv   
mwtpath                                                                         
/Volumes/COBOLT/MWT DISCARD/20100216X_JC_0s0x0s...      NaN   NaN    1.0  NaN   
/Volumes/COBOLT/MWT DISCARD/20100226X_JC_0s0x0s...      NaN   NaN    1.0  NaN   
/Volumes/COBOLT/MWT DISCARD/20100309X_JC_0s0x0s...      NaN   NaN    2.0  NaN   
/Volumes/COBOLT/MWT DISCARD/20100309X_JC_0s0x0s...      NaN   NaN    1.0  NaN   
/Volumes/COBOLT/MWT DISCARD/20110820X_SS_1500s3...      NaN   NaN    1.0  NaN   
...                                                     ...   ...    ...  ...   
/Volumes/COBOLT/MWT/20190418X_XX_100s30x10s10s_...      NaN   NaN    5.0  NaN   
/Volumes/COBOLT/MWT/20190418X_XX_100s30x10s10s_...      NaN   NaN    7.0  NaN   
/Volumes/COBOLT/MWT/20190418X_XX_100s30x10s10s_...      NaN   NaN    3.0  NaN   
/Volumes/COBOLT/MWT/20190418X_XX_100s30x10s10s_...      NaN   NaN    3.0  NaN   
/Volumes/COBOLT/MWT/20190418X_XX_100s30x10s10s_...      NaN   NaN    1.0  NaN   

                                                                            \
suffix                                             .dat .db .eps .fig .jpg   
mwtpath                                                                      
/Volumes/COBOLT/MWT DISCARD/20100216X_JC_0s0x0s...  NaN NaN  NaN  NaN  NaN   
/Volumes/COBOLT/MWT DISCARD/20100226X_JC_0s0x0s...  1.0 NaN  NaN  NaN  NaN   
/Volumes/COBOLT/MWT DISCARD/20100309X_JC_0s0x0s...  2.0 NaN  NaN  NaN  NaN   
/Volumes/COBOLT/MWT DISCARD/20100309X_JC_0s0x0s...  1.0 NaN  NaN  NaN  NaN   
/Volumes/COBOLT/MWT DISCARD/20110820X_SS_1500s3...  NaN NaN  NaN  NaN  NaN   
...                                                 ...  ..  ...  ...  ...   
/Volumes/COBOLT/MWT/20190418X_XX_100s30x10s10s_...  NaN NaN  NaN  NaN  NaN   
/Volumes/COBOLT/MWT/20190418X_XX_100s30x10s10s_...  NaN NaN  NaN  NaN  NaN   
/Volumes/COBOLT/MWT/20190418X_XX_100s30x10s10s_...  NaN NaN  NaN  NaN  NaN   
/Volumes/COBOLT/MWT/20190418X_XX_100s30x10s10s_...  NaN NaN  NaN  NaN  NaN   
/Volumes/COBOLT/MWT/20190418X_XX_100s30x10s10s_...  NaN NaN  NaN  NaN  NaN   

                                                                           \
suffix                                              .m .mat  .p .pdf .png   
mwtpath                                                                     
/Volumes/COBOLT/MWT DISCARD/20100216X_JC_0s0x0s... NaN  NaN NaN  NaN  1.0   
/Volumes/COBOLT/MWT DISCARD/20100226X_JC_0s0x0s... NaN  NaN NaN  NaN  1.0   
/Volumes/COBOLT/MWT DISCARD/20100309X_JC_0s0x0s... NaN  NaN NaN  NaN  1.0   
/Volumes/COBOLT/MWT DISCARD/20100309X_JC_0s0x0s... NaN  NaN NaN  NaN  1.0   
/Volumes/COBOLT/MWT DISCARD/20110820X_SS_1500s3... NaN  NaN NaN  NaN  1.0   
...                                                 ..  ...  ..  ...  ...   
/Volumes/COBOLT/MWT/20190418X_XX_100s30x10s10s_... NaN  1.0 NaN  NaN  2.0   
/Volumes/COBOLT/MWT/20190418X_XX_100s30x10s10s_... NaN  1.0 NaN  NaN  2.0   
/Volumes/COBOLT/MWT/20190418X_XX_100s30x10s10s_... NaN  1.0 NaN  NaN  2.0   
/Volumes/COBOLT/MWT/20190418X_XX_100s30x10s10s_... NaN  1.0 NaN  NaN  2.0   
/Volumes/COBOLT/MWT/20190418X_XX_100s30x10s10s_... NaN  1.0 NaN  NaN  2.0   

                                                                               \
suffix                                             .rasterData .rev .set .ssr   
mwtpath                                                                         
/Volumes/COBOLT/MWT DISCARD/20100216X_JC_0s0x0s...         NaN  NaN  NaN  NaN   
/Volumes/COBOLT/MWT DISCARD/20100226X_JC_0s0x0s...         NaN  NaN  NaN  NaN   
/Volumes/COBOLT/MWT DISCARD/20100309X_JC_0s0x0s...         NaN  1.0  NaN  NaN   
/Volumes/COBOLT/MWT DISCARD/20100309X_JC_0s0x0s...         NaN  1.0  NaN  NaN   
/Volumes/COBOLT/MWT DISCARD/20110820X_SS_1500s3...         NaN  NaN  1.0  NaN   
...                                                        ...  ...  ...  ...   
/Volumes/COBOLT

In [215]:
i = ~file_summary.iloc[:,-1].isna()
i.values

array([False, False, False, ..., False, False, False])

In [216]:
plates_with_trinity = file_summary.iloc[i.values,-1].copy()

In [217]:
plates_with_trinity.shape

(871,)

In [218]:
i = ~file_summary.iloc[:,-2].isna()
plates_with_swanlake = file_summary.iloc[i.values,-1].copy()
plates_with_swanlake.shape

(30,)

More plates with trinity than swanlake

In [219]:
plates_with_trinity.head()

mwtpath
/Volumes/COBOLT/MWT/20110906B_CL_100s30x60s10s_pilot/N2/20110906_113945    408.0
/Volumes/COBOLT/MWT/20110906B_CL_100s30x60s10s_pilot/N2/20110906_121214    495.0
/Volumes/COBOLT/MWT/20110906B_CL_100s30x60s10s_pilot/N2/20110906_124444    406.0
/Volumes/COBOLT/MWT/20110907B_SS_100s30x10s10s/NM1630/20110907_102613      369.0
/Volumes/COBOLT/MWT/20110907B_SS_100s30x10s10s/NM1630/20110907_103523      427.0
Name: (filepath, trinity.id.dat), dtype: float64

In [220]:
# get mwt paths
mwtpaths = pd.DataFrame(plates_with_trinity.index.values, columns=['paths'])
mwtpaths

paths
0    /Volumes/COBOLT/MWT/20110906B_CL_100s30x60s10s...
1    /Volumes/COBOLT/MWT/20110906B_CL_100s30x60s10s...
2    /Volumes/COBOLT/MWT/20110906B_CL_100s30x60s10s...
3    /Volumes/COBOLT/MWT/20110907B_SS_100s30x10s10s...
4    /Volumes/COBOLT/MWT/20110907B_SS_100s30x10s10s...
..                                                 ...
866  /Volumes/COBOLT/MWT/20170605X_CR_100s30x10s10s...
867  /Volumes/COBOLT/MWT/20170605X_CR_100s30x10s10s...
868  /Volumes/COBOLT/MWT/20170605X_CR_100s30x10s10s...
869  /Volumes/COBOLT/MWT/20170605X_CR_100s30x10s10s...
870  /Volumes/COBOLT/MWT/20170605X_CR_100s30x10s10s...

[871 rows x 1 columns]

In [221]:
# get which ones are alcohol and which ones are not
# split paths 
# MWTDB = pd.DataFrame()
MWTDB = mwtpaths['paths'].str.split(pat='/', expand=True)
MWTDB

0         1       2    3                                  4             5  \
0       Volumes  COBOLT  MWT   20110906B_CL_100s30x60s10s_pilot            N2   
1       Volumes  COBOLT  MWT   20110906B_CL_100s30x60s10s_pilot            N2   
2       Volumes  COBOLT  MWT   20110906B_CL_100s30x60s10s_pilot            N2   
3       Volumes  COBOLT  MWT         20110907B_SS_100s30x10s10s        NM1630   
4       Volumes  COBOLT  MWT         20110907B_SS_100s30x10s10s        NM1630   
..  ..      ...     ...  ...                                ...           ...   
866     Volumes  COBOLT  MWT  20170605X_CR_100s30x10s10s_TM1630  TM1630_400mM   
867     Volumes  COBOLT  MWT  20170605X_CR_100s30x10s10s_TM1630  TM1630_400mM   
868     Volumes  COBOLT  MWT  20170605X_CR_100s30x10s10s_TM1630  TM1630_400mM   
869     Volumes  COBOLT  MWT  20170605X_CR_100s30x10s10s_TM1630  TM1630_400mM   
870     Volumes  COBOLT  MWT  20170605X_CR_100s30x10s10s_TM1630  TM1630_400mM   

                   6  
0    20110906_113945  
1    20110906_121214  
2    20110906_124444  
3    20110907_102613  
4    20110907_103523  
..               ...  
866  20170605_134308  
867  20170605_141815  
868  20170605_144356  
869  20170605_152520  
870  20170605_154338  

[871 rows x 7 columns]

In [222]:
# see unique groups
pd.set_option('display.max_rows',100)
pd.DataFrame(MWTDB.iloc[:,5].value_counts().sort_values(ascending=False))

5
N2                            189
N2_400mM                      146
NM1630                         40
NM1968                         34
VC228                          18
VC228_400mM                    18
KP4                            16
KP4_400mM                      14
TM774_400mM                    12
VM3109_400mM                   12
VM3109                         12
NM1630_400mM                   12
N2_400mM_ExpTime60min          11
RB1231_400mM                   10
N2_400mM_NoFood                 9
BZ142                           9
TM5182_400mM                    8
TM2659                          8
TM474_400mM                     8
TM2179                          8
N2_400mM_ExpTime120min          8
TM1630                          8
TM474                           8
TM1630_400mM                    8
RM3389                          8
TM5182                          8
N2_400mM_ExpTime0min            8
RM3389_400mM                    8
TM2659_400mM                    8
N2_NoFood                       8
TM774                           7
DG1856                          7
TM8625                          7
TM2179_400mM                    7
TM8625_400mM                    7
JT748                           6
KP1097                          6
VC40862                         6
JT748_400mM                     6
VC40516                         6
VC40516_400mM                   6
VC40862_400mM                   6
PS2627                          6
N2_E3d12h400mM_R12h_T4d400mM    6
KP1097_400mM                    6
TM4724                          6
PS2627_400mM                    6
TM4724_400mM                    6
NM1968_Paxillin1pM              5
NM1968_Paxillin10pM             5
N2_Paxillin100pM                5
N2_400mM_ExpTime90min           4
RB989_400mM                     4
VC1416                          4
RB989                           4
N2_TsfLiquidTime120to180        4
N2_TsfLiquidTime30to90          4
VG302_400mM                     4
VG302                           4
VC1416_400mM                    4
SG1_400mM                       4
CX3940                          4
SG1                             4
DA465                           3
NM1968_Paxillin1mM              3
NM1968_Paxillin10mM             3
CX3940_Paxillin100pM            2
NM1968_Paxillin10uM             2
NM1968_Paxillin1uM              2
NM1968_Paxillin100pM            2
CX3940_Paxillin1pM              1
RB982_400mM                     1
VG202_400mM_NoLid               1
MT2426                          1

In [223]:
# take just N2 and N2 400mM to train
i = MWTDB.iloc[:,5].str.match(pat='^N2$') | MWTDB.iloc[:,5].str.match(pat='^N2_400mM$')
# check to see if plate names are unique
if MWTDB.loc[i,6].unique().shape == np.sum(i):
    print('plate names unique')
mwtpaths_target = MWTDB.loc[i,:].copy()
mwtpaths_target['mwtpath'] = mwtpaths[i]


plate names unique


# get paths to trinity data from file database


In [224]:
#load file database csv
df = pd.read_csv(os.path.join(dir_save,'path_allfiles_MWTDB.csv'))
df

Unnamed: 0        dbname                        expname groupname  \
0                 0           MWT  20100201X_JC_0s0x0s0s_EE01013    N2_EE0   
1                 1           MWT  20100201X_JC_0s0x0s0s_EE01013    N2_EE0   
2                 2           MWT  20100201X_JC_0s0x0s0s_EE01013    N2_EE0   
3                 3           MWT  20100201X_JC_0s0x0s0s_EE01013    N2_EE0   
4                 4           MWT  20100201X_JC_0s0x0s0s_EE01013    N2_EE0   
...             ...           ...                            ...       ...   
1318719     1318719  MWT to check     20140819C_SM_100s30x10s10s   ungroup   
1318720     1318720  MWT to check     20140819C_SM_100s30x10s10s   ungroup   
1318721     1318721  MWT to check     20140819C_SM_100s30x10s10s   ungroup   
1318722     1318722  MWT to check     20140819C_SM_100s30x10s10s   ungroup   
1318723     1318723  MWT to check     20140819C_SM_100s30x10s10s   ungroup   

                 mwtname                                           filename  \
0        20100201_131356                                        0%_EtOH.dat   
1        20100201_131356                                        0%_EtOH.png   
2        20100201_131356                                    0%_EtOH.summary   
3        20100201_131356                                   0%_EtOH.evan.dat   
4        20100201_131356                          0%_EtOH.drunkposture2.dat   
...                  ...                                                ...   
1318719  20140819_192732           N2_5x3_t96h20C_100s30x10s10s_C0109dd.set   
1318720  20140819_192732       N2_5x3_t96h20C_100s30x10s10s_C0109dd.summary   
1318721  20140819_192732  N2_5x3_t96h20C_100s30x10s10s_C0109dd_00000k.blobs   
1318722  20140819_192732  N2_5x3_t96h20C_100s30x10s10s_C0109dd_00001k.blobs   
1318723  20140819_192732  N2_5x3_t96h20C_100s30x10s10s_C0109dd.drunkpost...   

                                                      path  \
0        /Volumes/COBOLT/MWT/20100201X_JC_0s0x0s0s_EE01...   
1        /Volumes/COBOLT/MWT/20100201X_JC_0s0x0s0s_EE01...   
2        /Volumes/COBOLT/MWT/20100201X_JC_0s0x0s0s_EE01...   
3        /Volumes/COBOLT/MWT/20100201X_JC_0s0x0s0s_EE01...   
4        /Volumes/COBOLT/MWT/20100201X_JC_0s0x0s0s_EE01...   
...                                                    ...   
1318719  /Volumes/COBOLT/MWT to check/20140819C_SM_100s...   
1318720  /Volumes/COBOLT/MWT to check/20140819C_SM_100s...   
1318721  /Volumes/COBOLT/MWT to check/20140819C_SM_100s...   
1318722  /Volumes/COBOLT/MWT to check/20140819C_SM_100s...   
1318723  /Volumes/COBOLT/MWT to check/20140819C_SM_100s...   

                                                     fname       ext  \
0                                                  0%_EtOH      .dat   
1                                                  0%_EtOH      .png   
2                                                  0%_EtOH  .summary   
3                                             0%_EtOH.evan      .dat   
4                                    0%_EtOH.drunkposture2      .dat   
...                                                    ...       ...   
1318719               N2_5x3_t96h20C_100s30x10s10s_C0109dd      .set   
1318720               N2_5x3_t96h20C_100s30x10s10s_C0109dd  .summary   
1318721        N2_5x3_t96h20C_100s30x10s10s_C0109dd_00000k    .blobs   
1318722        N2_5x3_t96h20C_100s30x10s10s_C0109dd_00001k    .blobs   
1318723  N2_5x3_t96h20C_100s30x10s10s_C0109dd.drunkpost...      .dat   

               analysis_tag  
0                        []  
1                        []  
2                        []  
3                 ['.evan']  
4        ['.drunkposture2']  
...                     ...  
1318719                  []  
1318720                  []  
1318721                  []  
1318722                  []  
1318723  ['.drunkposture2']  

[1318724 rows x 10 columns]

In [225]:
i = df['mwtname'].isin(mwtpaths_target.iloc[:,6].values)


In [226]:
sum(i)/len(i)

0.1718486961638675

In [227]:
# reduce data frame
df = df.loc[i,:].copy()
df.shape

(226621, 10)

In [228]:
i = df['filename'].str.contains(r'trinity[.]\d{1,}(.)dat$')

/Users/connylin/opt/anaconda3/lib/python3.7/site-packages/pandas/core/strings.py:1954: UserWarning: This pattern has match groups. To actually get the groups, use str.extract.
  return func(self, *args, **kwargs)


In [229]:
# reduce again to only file name containing trinity
df = df.loc[i,:].copy()
df.shape

(216261, 10)

In [230]:
# print some to confirm
for f in df['filename'].iloc[:10]:
    print(f)

N2_5x3_f92h20C_100s30x60s10s_B0902bb.trinity.00001.dat
N2_5x3_f92h20C_100s30x60s10s_B0902bb.trinity.00002.dat
N2_5x3_f92h20C_100s30x60s10s_B0902bb.trinity.04098.dat
N2_5x3_f92h20C_100s30x60s10s_B0902bb.trinity.00004.dat
N2_5x3_f92h20C_100s30x60s10s_B0902bb.trinity.00005.dat
N2_5x3_f92h20C_100s30x60s10s_B0902bb.trinity.04101.dat
N2_5x3_f92h20C_100s30x60s10s_B0902bb.trinity.00006.dat
N2_5x3_f92h20C_100s30x60s10s_B0902bb.trinity.04103.dat
N2_5x3_f92h20C_100s30x60s10s_B0902bb.trinity.00008.dat
N2_5x3_f92h20C_100s30x60s10s_B0902bb.trinity.00009.dat


In [231]:
# group by 0mM and 400mM to see how many data we have per group
df['groupname'].value_counts()

N2             120524
N2_400mM        92886
TM774_400mM      2799
VC40862            52
Name: groupname, dtype: int64

In [232]:
# found some dirty data. delete those
i = df['groupname'].isin(['TM774_400mM','VC40862'])
df.drop(df.index[i], axis=0, inplace=True)

In [233]:
df.shape

(213410, 10)

In [234]:
# group by 0mM and 400mM to see how many data we have per group
df['groupname'].value_counts()

N2          120524
N2_400mM     92886
Name: groupname, dtype: int64

Now the db is clean

the samples are a little unbalanced but not too bad.

In [235]:
# store df in more permanent variable name
fileDB_target = df.copy()
del df

In [269]:
# get worm id
worm_number = fileDB_target['fname'].str.extract(r'(?<=trinity[.])(\d{1,})')
worm_number = worm_number.astype('int32').values

In [334]:
fileDB_target['worm_id'] = worm_number

# take a look at trinity data

In [237]:
# sample data to see how to import the data
pfile = fileDB_target['path'].iloc[200]
pfile

'/Volumes/COBOLT/MWT/20110906B_CL_100s30x60s10s_pilot/N2/20110906_113945/N2_5x3_f92h20C_100s30x60s10s_B0902bb.trinity.05483.dat'

In [238]:
df = pd.read_csv(pfile, delim_whitespace=True, header=None)
df
# if the file has worm id can also concat directly from bash

0    1    2       3    4    5    6    7        8        9       10  \
0    1709.275  1.0  1.0     NaN  0.0  1.0  0.0  0.0  39.0142  18.2930  0.1093   
1    1709.360  1.0  1.0  0.1072  0.0  1.0  0.0  0.0  39.0191  18.3024  0.1112   
2    1709.446  1.0  1.0  0.0959  0.0  1.0  0.0  0.0  39.0225  18.3093  0.1094   
3    1709.518  1.0  1.0  0.0956  0.0  1.0  0.0  0.0  39.0256  18.3159  0.1106   
4    1709.592  1.0  1.0  0.0713  0.0  1.0  0.0  0.0  39.0286  18.3218  0.1120   
..        ...  ...  ...     ...  ...  ...  ...  ...      ...      ...     ...   
489  1739.672  1.0  1.0  0.1655  0.0  1.0  0.0  0.0  40.4880  20.4281  0.1121   
490  1739.711  1.0  1.0  0.2041  0.0  0.0  0.0  0.0  40.4941  20.4343  0.1141   
491  1739.748  1.0  1.0  0.2142  0.0  0.0  0.0  0.0  40.5004  20.4388  0.1144   
492  1739.811  1.0  1.0     NaN  0.0  0.0  0.0  0.0  40.5106  20.4479  0.1130   
493  1739.877  1.0  1.0     NaN  0.0  0.0  0.0  0.0  40.5225  20.4573  0.1108   

         11        12   13     14    15    16      17  
0    1.2487  0.173502  NaN  0.208  20.1  35.5     NaN  
1    1.2643  0.175689  5.2  0.202  19.4  37.2  0.0012  
2    1.2437  0.174231  4.5  0.207  22.4  39.0  0.0030  
3    1.2483  0.174231  4.6  0.201  23.3  38.9  0.0017  
4    1.2594  0.175689  4.0  0.199  22.8  38.6  0.0054  
..      ...       ...  ...    ...   ...   ...     ...  
489  1.3017  0.177147  9.1  0.365  43.5  38.5  0.0289  
490  1.2576  0.175689  7.4  0.359  51.0  40.9  0.0312  
491  1.2671  0.176418  6.6  0.330  46.7  41.1  0.0301  
492  1.2656  0.174960  NaN  0.337  46.6  41.4     NaN  
493  1.2987  0.175689  NaN  0.325  49.6  41.6     NaN  

[494 rows x 18 columns]

In [239]:
# load file summary
filename = 'legend_trinity_worm.pickle'
with open(os.path.join(dir_save, filename),'rb') as f:
    v = pickle.load(f)

In [245]:
df.columns = v['name'].values

In [250]:
# only keep variables that are informative
col_ind_keep = [0,3,5,6,8,9,10,11,12,13,14,15,16,17]
df = pd.read_csv(pfile, usecols=col_ind_keep, delim_whitespace=True, 
                 header=None, names=v.loc[col_ind_keep,'name'].values)

time   speed  bias  tap    loc_x    loc_y  morphwidth  midline  \
0    1709.275     NaN   1.0  0.0  39.0142  18.2930      0.1093   1.2487   
1    1709.360  0.1072   1.0  0.0  39.0191  18.3024      0.1112   1.2643   
2    1709.446  0.0959   1.0  0.0  39.0225  18.3093      0.1094   1.2437   
3    1709.518  0.0956   1.0  0.0  39.0256  18.3159      0.1106   1.2483   
4    1709.592  0.0713   1.0  0.0  39.0286  18.3218      0.1120   1.2594   
..        ...     ...   ...  ...      ...      ...         ...      ...   
489  1739.672  0.1655   1.0  0.0  40.4880  20.4281      0.1121   1.3017   
490  1739.711  0.2041   0.0  0.0  40.4941  20.4343      0.1141   1.2576   
491  1739.748  0.2142   0.0  0.0  40.5004  20.4388      0.1144   1.2671   
492  1739.811     NaN   0.0  0.0  40.5106  20.4479      0.1130   1.2656   
493  1739.877     NaN   0.0  0.0  40.5225  20.4573      0.1108   1.2987   

         area  angular  aspect  kink  curve    crab  
0    0.173502      NaN   0.208  20.1   35.5     NaN  
1    0.175689      5.2   0.202  19.4   37.2  0.0012  
2    0.174231      4.5   0.207  22.4   39.0  0.0030  
3    0.174231      4.6   0.201  23.3   38.9  0.0017  
4    0.175689      4.0   0.199  22.8   38.6  0.0054  
..        ...      ...     ...   ...    ...     ...  
489  0.177147      9.1   0.365  43.5   38.5  0.0289  
490  0.175689      7.4   0.359  51.0   40.9  0.0312  
491  0.176418      6.6   0.330  46.7   41.1  0.0301  
492  0.174960      NaN   0.337  46.6   41.4     NaN  
493  0.175689      NaN   0.325  49.6   41.6     NaN  

[494 rows x 14 columns]

## Get row numbers

Research:
* https://stackoverflow.com/questions/45091398/pandas-combine-multiple-csvs-and-output-as-one-large-file

In [ ]:
# count rows for each file
# create worm id with that many rows

In [270]:
fileDB_target.shape

(213410, 10)

In [332]:
# get row numbers for all of the files
# declare empty array to store the row numbers
wormid_array = np.empty((fileDB_target.shape[0],1),dtype='int')
# loop through each files 
for i, p in enumerate(fileDB_target['path'], 0):
    # report process every 100 files
    if (i%100==0):
        print(f'processing #{i} files', end='\r')
    # create custom bash command
    bashCommand = "cat "+p+" | wc -l"
    # call bash
    output = subprocess.check_output(bashCommand, shell=True)
    # take out digits from the byte output
    wormid_array[i] = np.array(output,dtype='int')
wormid_array

array([[ 527],
       [ 915],
       [9648],
       ...,
       [ 155],
       [1649],
       [ 714]])

In [335]:
fileDB_target['row_number'] = wormid_array

In [337]:
fileDB_target

Unnamed: 0 dbname                            expname groupname  \
39135         39135    MWT   20110906B_CL_100s30x60s10s_pilot        N2   
39136         39136    MWT   20110906B_CL_100s30x60s10s_pilot        N2   
39137         39137    MWT   20110906B_CL_100s30x60s10s_pilot        N2   
39138         39138    MWT   20110906B_CL_100s30x60s10s_pilot        N2   
39139         39139    MWT   20110906B_CL_100s30x60s10s_pilot        N2   
...             ...    ...                                ...       ...   
1307197     1307197    MWT  20170605X_CR_100s30x10s10s_TM1630  N2_400mM   
1307198     1307198    MWT  20170605X_CR_100s30x10s10s_TM1630  N2_400mM   
1307199     1307199    MWT  20170605X_CR_100s30x10s10s_TM1630  N2_400mM   
1307200     1307200    MWT  20170605X_CR_100s30x10s10s_TM1630  N2_400mM   
1307201     1307201    MWT  20170605X_CR_100s30x10s10s_TM1630  N2_400mM   

                 mwtname                                           filename  \
39135    20110906_113945  N2_5x3_f92h20C_100s30x60s10s_B0902bb.trinity.0...   
39136    20110906_113945  N2_5x3_f92h20C_100s30x60s10s_B0902bb.trinity.0...   
39137    20110906_113945  N2_5x3_f92h20C_100s30x60s10s_B0902bb.trinity.0...   
39138    20110906_113945  N2_5x3_f92h20C_100s30x60s10s_B0902bb.trinity.0...   
39139    20110906_113945  N2_5x3_f92h20C_100s30x60s10s_B0902bb.trinity.0...   
...                  ...                                                ...   
1307197  20170605_153818  C8_tm1630_30x0_t96h20C_100s30x10s10s_C0605_CMR...   
1307198  20170605_153818  C8_tm1630_30x0_t96h20C_100s30x10s10s_C0605_CMR...   
1307199  20170605_153818  C8_tm1630_30x0_t96h20C_100s30x10s10s_C0605_CMR...   
1307200  20170605_153818  C8_tm1630_30x0_t96h20C_100s30x10s10s_C0605_CMR...   
1307201  20170605_153818  C8_tm1630_30x0_t96h20C_100s30x10s10s_C0605_CMR...   

                                                      path  \
39135    /Volumes/COBOLT/MWT/20110906B_CL_100s30x60s10s...   
39136    /Volumes/COBOLT/MWT/20110906B_CL_100s30x60s10s...   
39137    /Volumes/COBOLT/MWT/20110906B_CL_100s30x60s10s...   
39138    /Volumes/COBOLT/MWT/20110906B_CL_100s30x60s10s...   
39139    /Volumes/COBOLT/MWT/20110906B_CL_100s30x60s10s...   
...                                                    ...   
1307197  /Volumes/COBOLT/MWT/20170605X_CR_100s30x10s10s...   
1307198  /Volumes/COBOLT/MWT/20170605X_CR_100s30x10s10s...   
1307199  /Volumes/COBOLT/MWT/20170605X_CR_100s30x10s10s...   
1307200  /Volumes/COBOLT/MWT/20170605X_CR_100s30x10s10s...   
1307201  /Volumes/COBOLT/MWT/20170605X_CR_100s30x10s10s...   

                                                     fname   ext  \
39135    N2_5x3_f92h20C_100s30x60s10s_B0902bb.trinity.0...  .dat   
39136    N2_5x3_f92h20C_100s30x60s10s_B0902bb.trinity.0...  .dat   
39137    N2_5x3_f92h20C_100s30x60s10s_B0902bb.trinity.0...  .dat   
39138    N2_5x3_f92h20C_100s30x60s10s_B0902bb.trinity.0...  .dat   
39139    N2_5x3_f92h20C_100s30x60s10s_B0902bb.trinity.0...  .dat   
...                                                    ...   ...   
1307197  C8_tm1630_30x0_t96h20C_100s30x10s10s_C0605_CMR...  .dat   
1307198  C8_tm1630_30x0_t96h20C_100s30x10s10s_C0605_CMR...  .dat   
1307199  C8_tm1630_30x0_t96h20C_100s30x10s10s_C0605_CMR...  .dat   
1307200  C8_tm1630_30x0_t96h20C_100s30x10s10s_C0605_CMR...  .dat   
1307201  C8_tm1630_30x0_t96h20C_100s30x10s10s_C0605_CMR...  .dat   

                  analysis_tag  worm_id  row_number  
39135    ['.00001' '.trinity']        1         527  
39136    ['.00002' '.trinity']        2         915  
39137    ['.04098' '.trinity']     4098        9648  
39138    ['.00004' '.trinity']        4        1611  
39139    ['.00005' '.trinity']        5         707  
...                        ...      ...         ...  
1307197  ['.02552' '.trinity']     2552        1186  
1307198  ['.00505' '.trinity']      505         477  
1307199  ['.02041' '.trinity']     2041         155  
1307200  ['.01531' '.trinity']     1531        1649  
130

In [117]:
# create group name
groupname_array = np.empty([fileDB_target.shape[0],1], dtype='int')

groupname_array[np.array(fileDB_target['groupname']=='N2')]=0
groupname_array[np.array(fileDB_target['groupname']=='N2_400mM')]=1
fileDB_target['ethanol'] = groupname_array

Unnamed: 0 dbname                            expname groupname  \
39135         39135    MWT   20110906B_CL_100s30x60s10s_pilot        N2   
39136         39136    MWT   20110906B_CL_100s30x60s10s_pilot        N2   
39137         39137    MWT   20110906B_CL_100s30x60s10s_pilot        N2   
39138         39138    MWT   20110906B_CL_100s30x60s10s_pilot        N2   
39139         39139    MWT   20110906B_CL_100s30x60s10s_pilot        N2   
...             ...    ...                                ...       ...   
1307197     1307197    MWT  20170605X_CR_100s30x10s10s_TM1630  N2_400mM   
1307198     1307198    MWT  20170605X_CR_100s30x10s10s_TM1630  N2_400mM   
1307199     1307199    MWT  20170605X_CR_100s30x10s10s_TM1630  N2_400mM   
1307200     1307200    MWT  20170605X_CR_100s30x10s10s_TM1630  N2_400mM   
1307201     1307201    MWT  20170605X_CR_100s30x10s10s_TM1630  N2_400mM   

                 mwtname                                           filename  \
39135    20110906_113945  N2_5x3_f92h20C_100s30x60s10s_B0902bb.trinity.0...   
39136    20110906_113945  N2_5x3_f92h20C_100s30x60s10s_B0902bb.trinity.0...   
39137    20110906_113945  N2_5x3_f92h20C_100s30x60s10s_B0902bb.trinity.0...   
39138    20110906_113945  N2_5x3_f92h20C_100s30x60s10s_B0902bb.trinity.0...   
39139    20110906_113945  N2_5x3_f92h20C_100s30x60s10s_B0902bb.trinity.0...   
...                  ...                                                ...   
1307197  20170605_153818  C8_tm1630_30x0_t96h20C_100s30x10s10s_C0605_CMR...   
1307198  20170605_153818  C8_tm1630_30x0_t96h20C_100s30x10s10s_C0605_CMR...   
1307199  20170605_153818  C8_tm1630_30x0_t96h20C_100s30x10s10s_C0605_CMR...   
1307200  20170605_153818  C8_tm1630_30x0_t96h20C_100s30x10s10s_C0605_CMR...   
1307201  20170605_153818  C8_tm1630_30x0_t96h20C_100s30x10s10s_C0605_CMR...   

                                                      path  \
39135    /Volumes/COBOLT/MWT/20110906B_CL_100s30x60s10s...   
39136    /Volumes/COBOLT/MWT/20110906B_CL_100s30x60s10s...   
39137    /Volumes/COBOLT/MWT/20110906B_CL_100s30x60s10s...   
39138    /Volumes/COBOLT/MWT/20110906B_CL_100s30x60s10s...   
39139    /Volumes/COBOLT/MWT/20110906B_CL_100s30x60s10s...   
...                                                    ...   
1307197  /Volumes/COBOLT/MWT/20170605X_CR_100s30x10s10s...   
1307198  /Volumes/COBOLT/MWT/20170605X_CR_100s30x10s10s...   
1307199  /Volumes/COBOLT/MWT/20170605X_CR_100s30x10s10s...   
1307200  /Volumes/COBOLT/MWT/20170605X_CR_100s30x10s10s...   
1307201  /Volumes/COBOLT/MWT/20170605X_CR_100s30x10s10s...   

                                                     fname   ext  \
39135    N2_5x3_f92h20C_100s30x60s10s_B0902bb.trinity.0...  .dat   
39136    N2_5x3_f92h20C_100s30x60s10s_B0902bb.trinity.0...  .dat   
39137    N2_5x3_f92h20C_100s30x60s10s_B0902bb.trinity.0...  .dat   
39138    N2_5x3_f92h20C_100s30x60s10s_B0902bb.trinity.0...  .dat   
39139    N2_5x3_f92h20C_100s30x60s10s_B0902bb.trinity.0...  .dat   
...                                                    ...   ...   
1307197  C8_tm1630_30x0_t96h20C_100s30x10s10s_C0605_CMR...  .dat   
1307198  C8_tm1630_30x0_t96h20C_100s30x10s10s_C0605_CMR...  .dat   
1307199  C8_tm1630_30x0_t96h20C_100s30x10s10s_C0605_CMR...  .dat   
1307200  C8_tm1630_30x0_t96h20C_100s30x10s10s_C0605_CMR...  .dat   
1307201  C8_tm1630_30x0_t96h20C_100s30x10s10s_C0605_CMR...  .dat   

                  analysis_tag  worm_id  row_number  ethanol  
39135    ['.00001' '.trinity']        1         527        0  
39136    ['.00002' '.trinity']        2         915        0  
39137    ['.04098' '.trinity']     4098        9648        0  
39138    ['.00004' '.trinity']        4        1611        0  
39139    ['.00005' '.trinity']        5         707        0  
...                        ...      ...         ...      ...  
1307197  ['.02552' '.trinity']     2552        1186        1  
1307198  ['.00505' '.trinity']      505         477        1  
1307199  ['.02041' '.trinity']

In [118]:
# save in pickle
filename = 'fileinfo_trinity_N2400mM.pickle'
pickle.dump(fileDB_target,open(os.path.join(dir_save, filename),'wb'))

https://stackoverflow.com/questions/20716812/saving-and-loading-multiple-objects-in-pickle-file
```
import pickle

file = open('test.pkl','wb')
obj_1 = ['test_1', {'ability', 'mobility'}]
obj_2 = ['test_2', {'ability', 'mobility'}]
obj_3 = ['test_3', {'ability', 'mobility'}]

pickle.dump(obj_1, file)
pickle.dump(obj_2, file)
pickle.dump(obj_3, file)

file.close()

file = open('test.pkl', 'rb')
obj_1 = pickle.load(file)
obj_2 = pickle.load(file)
obj_3 = pickle.load(file)
print(obj_1)
print(obj_2)
print(obj_3)
file.close()
```

In [11]:
# load pickle
filename = 'fileinfo_trinity_N2400mM.pickle'
fileDB_target = pickle.load(open(os.path.join(dir_save, filename),'rb'))

In [119]:
fileDB_target

Unnamed: 0 dbname                            expname groupname  \
39135         39135    MWT   20110906B_CL_100s30x60s10s_pilot        N2   
39136         39136    MWT   20110906B_CL_100s30x60s10s_pilot        N2   
39137         39137    MWT   20110906B_CL_100s30x60s10s_pilot        N2   
39138         39138    MWT   20110906B_CL_100s30x60s10s_pilot        N2   
39139         39139    MWT   20110906B_CL_100s30x60s10s_pilot        N2   
...             ...    ...                                ...       ...   
1307197     1307197    MWT  20170605X_CR_100s30x10s10s_TM1630  N2_400mM   
1307198     1307198    MWT  20170605X_CR_100s30x10s10s_TM1630  N2_400mM   
1307199     1307199    MWT  20170605X_CR_100s30x10s10s_TM1630  N2_400mM   
1307200     1307200    MWT  20170605X_CR_100s30x10s10s_TM1630  N2_400mM   
1307201     1307201    MWT  20170605X_CR_100s30x10s10s_TM1630  N2_400mM   

                 mwtname                                           filename  \
39135    20110906_113945  N2_5x3_f92h20C_100s30x60s10s_B0902bb.trinity.0...   
39136    20110906_113945  N2_5x3_f92h20C_100s30x60s10s_B0902bb.trinity.0...   
39137    20110906_113945  N2_5x3_f92h20C_100s30x60s10s_B0902bb.trinity.0...   
39138    20110906_113945  N2_5x3_f92h20C_100s30x60s10s_B0902bb.trinity.0...   
39139    20110906_113945  N2_5x3_f92h20C_100s30x60s10s_B0902bb.trinity.0...   
...                  ...                                                ...   
1307197  20170605_153818  C8_tm1630_30x0_t96h20C_100s30x10s10s_C0605_CMR...   
1307198  20170605_153818  C8_tm1630_30x0_t96h20C_100s30x10s10s_C0605_CMR...   
1307199  20170605_153818  C8_tm1630_30x0_t96h20C_100s30x10s10s_C0605_CMR...   
1307200  20170605_153818  C8_tm1630_30x0_t96h20C_100s30x10s10s_C0605_CMR...   
1307201  20170605_153818  C8_tm1630_30x0_t96h20C_100s30x10s10s_C0605_CMR...   

                                                      path  \
39135    /Volumes/COBOLT/MWT/20110906B_CL_100s30x60s10s...   
39136    /Volumes/COBOLT/MWT/20110906B_CL_100s30x60s10s...   
39137    /Volumes/COBOLT/MWT/20110906B_CL_100s30x60s10s...   
39138    /Volumes/COBOLT/MWT/20110906B_CL_100s30x60s10s...   
39139    /Volumes/COBOLT/MWT/20110906B_CL_100s30x60s10s...   
...                                                    ...   
1307197  /Volumes/COBOLT/MWT/20170605X_CR_100s30x10s10s...   
1307198  /Volumes/COBOLT/MWT/20170605X_CR_100s30x10s10s...   
1307199  /Volumes/COBOLT/MWT/20170605X_CR_100s30x10s10s...   
1307200  /Volumes/COBOLT/MWT/20170605X_CR_100s30x10s10s...   
1307201  /Volumes/COBOLT/MWT/20170605X_CR_100s30x10s10s...   

                                                     fname   ext  \
39135    N2_5x3_f92h20C_100s30x60s10s_B0902bb.trinity.0...  .dat   
39136    N2_5x3_f92h20C_100s30x60s10s_B0902bb.trinity.0...  .dat   
39137    N2_5x3_f92h20C_100s30x60s10s_B0902bb.trinity.0...  .dat   
39138    N2_5x3_f92h20C_100s30x60s10s_B0902bb.trinity.0...  .dat   
39139    N2_5x3_f92h20C_100s30x60s10s_B0902bb.trinity.0...  .dat   
...                                                    ...   ...   
1307197  C8_tm1630_30x0_t96h20C_100s30x10s10s_C0605_CMR...  .dat   
1307198  C8_tm1630_30x0_t96h20C_100s30x10s10s_C0605_CMR...  .dat   
1307199  C8_tm1630_30x0_t96h20C_100s30x10s10s_C0605_CMR...  .dat   
1307200  C8_tm1630_30x0_t96h20C_100s30x10s10s_C0605_CMR...  .dat   
1307201  C8_tm1630_30x0_t96h20C_100s30x10s10s_C0605_CMR...  .dat   

                  analysis_tag  worm_id  row_number  ethanol  
39135    ['.00001' '.trinity']        1         527        0  
39136    ['.00002' '.trinity']        2         915        0  
39137    ['.04098' '.trinity']     4098        9648        0  
39138    ['.00004' '.trinity']        4        1611        0  
39139    ['.00005' '.trinity']        5         707        0  
...                        ...      ...         ...      ...  
1307197  ['.02552' '.trinity']     2552        1186        1  
1307198  ['.00505' '.trinity']      505         477        1  
1307199  ['.02041' '.trinity']

In [6]:
# look at estimates
fileDB_target['row_number'].sum()

202718463

202,718,463. 202 Million rows.

# Take 1000 worms from each group to test alogorithms

In [32]:
# get sample
n_sample = 10
groupnames = ['N2','N2_400mM']
groupname_target = 'N2'

In [33]:
# separate by group
i = fileDB_target['groupname']==groupname_target
df_target = fileDB_target.iloc[i.values,:].copy()

In [34]:
df_target.shape

(120524, 12)

In [ ]:
df_sample = df_target.sample(n_sample, random_state=318).copy()

In [49]:
# get row total
row_total = df_sample['row_number'].sum()
# declare array
worm_id_array = np.empty([row_total,1],dtype='int')
worm_id_list = df_sample['worm_id'].values
rows_per_worm_list = df_sample['row_number'].values
i_row_previous = 0
for worm_id, rows in zip(worm_id_list, rows_per_worm_list):
    # get start row position
    i_start_row = i_row_previous
    # get end row position
    i_end_row = i_row_previous+rows-1
    print(i_row_previous, worm_id, rows, i_start_row, i_end_row)
    # get worm id array x number of rows
    worm_id_array[i_start_row:i_end_row+1,0] = np.tile(worm_id, rows)
    # create next starting position
    i_row_previous = rows+i_row_previous

0 12099 1630 0 1629
1630 3187 236 1630 1865
1866 17056 262 1866 2127
2128 402 267 2128 2394
2395 4610 798 2395 3192
3193 2895 788 3193 3980
3981 1327 257 3981 4237
4238 23505 113 4238 4350
4351 6735 470 4351 4820
4821 18168 1095 4821 5915


array([12099,  3187, 17056,   402,  4610,  2895,  1327, 23505,  6735,
       18168], dtype=int32)

https://www.datacamp.com/community/tutorials/shell-commands-data-scientist

`cat file_1.csv file_2.csv > target_file.csv`


In [95]:
# create bash command components
bashcommand_prefix = 'cat'
bashcommand_sourcepaths = ''
bashcommand_space = ' '
bashcommand_link = '>'
bashcommand_outputpath = '"'+os.path.join(dir_save, 'test_cat_csv.csv')+'"'

In [96]:
# link file paths
for p in df_sample['path'].values:
    p = '"'+p+'"'
    bashcommand_sourcepaths += p
    bashcommand_sourcepaths += ' '
print(bashcommand_sourcepaths)

"/Volumes/COBOLT/MWT/20120426C_DA_100s30x45s10s/N2/20120426_163031/N2_5x3_t90h20C_100s30x45s10s_C0422ab.trinity.12099.dat" "/Volumes/COBOLT/MWT/20140118B_SJ_100s30x60s10s_neurexin/N2/20140118_133159/N2_5x3_t96h20C_100s30x10s10s_C0109ed.trinity.03187.dat" "/Volumes/COBOLT/MWT/20120418C_CL_100s30x60s10s/N2/20120418_183401/N2_8x3_f98h20C_100s30x60s60s_C0414ab.trinity.17056.dat" "/Volumes/COBOLT/MWT/20170214X_XX_100s30x10s10s_KP1097/N2/20170214_111855/A5_15x0_t96h20C_100s30x10s10s_N2.trinity.00402.dat" "/Volumes/COBOLT/MWT/20120408C_DA_100s30x60s10s/N2/20120408_181122/N2_5x3_f92h20C_100s30x60s10s_C0404ab.trinity.04610.dat" "/Volumes/COBOLT/MWT/20111203C_CL_100s30x60s10s_discard/N2/20111203_160048/N2_5x24_t96h20C_100s30x60s60s_C1130ab.trinity.02895.dat" "/Volumes/COBOLT/MWT/20170504X_XX_100s30x10s10s_tm474/N2/20170504_132835/A6_15x0_t96h20C_100s30x10s10s_TM474.trinity.01327.dat" "/Volumes/COBOLT/MWT/20120426C_DA_100s30x45s10s/N2/20120426_163031/N2_5x3_t90h20C_100s30x45s10s_C0422ab.trinity.2

In [97]:
# final bash command
bashcommand = bashcommand_prefix + bashcommand_space + \
            bashcommand_sourcepaths + \
            bashcommand_link + bashcommand_space + \
            bashcommand_outputpath
bashcommand

'cat "/Volumes/COBOLT/MWT/20120426C_DA_100s30x45s10s/N2/20120426_163031/N2_5x3_t90h20C_100s30x45s10s_C0422ab.trinity.12099.dat" "/Volumes/COBOLT/MWT/20140118B_SJ_100s30x60s10s_neurexin/N2/20140118_133159/N2_5x3_t96h20C_100s30x10s10s_C0109ed.trinity.03187.dat" "/Volumes/COBOLT/MWT/20120418C_CL_100s30x60s10s/N2/20120418_183401/N2_8x3_f98h20C_100s30x60s60s_C0414ab.trinity.17056.dat" "/Volumes/COBOLT/MWT/20170214X_XX_100s30x10s10s_KP1097/N2/20170214_111855/A5_15x0_t96h20C_100s30x10s10s_N2.trinity.00402.dat" "/Volumes/COBOLT/MWT/20120408C_DA_100s30x60s10s/N2/20120408_181122/N2_5x3_f92h20C_100s30x60s10s_C0404ab.trinity.04610.dat" "/Volumes/COBOLT/MWT/20111203C_CL_100s30x60s10s_discard/N2/20111203_160048/N2_5x24_t96h20C_100s30x60s60s_C1130ab.trinity.02895.dat" "/Volumes/COBOLT/MWT/20170504X_XX_100s30x10s10s_tm474/N2/20170504_132835/A6_15x0_t96h20C_100s30x10s10s_TM474.trinity.01327.dat" "/Volumes/COBOLT/MWT/20120426C_DA_100s30x45s10s/N2/20120426_163031/N2_5x3_t90h20C_100s30x45s10s_C0422ab.trin

In [98]:
os.system(bashcommand)

0

The resulting csv is 677KB. Estimate size if get all files

In [84]:
d_size_est = fileDB_target.shape[0] / 10 * 677 / 1000**2
print(f'estimated size of final csv = {d_size_est:.1f} GB')

estimated size of final csv = 14.4 GB


## (not done) try to reduce this by taking only the columns of interest

https://www.datacamp.com/community/tutorials/shell-commands-data-scientist

```
cut -d "," -f 2 adult.csv | head -3
workclass
State-gov
Self-emp-not-inc
```


```
cat <filenames> | awk -F' ' '{print $1" "$4" "$2}' > filename.csv
```

In [92]:
# try with one file
# create bash command components
bashcommand_columns = "cut -d ' ' -f 0,3,5,6,8,9,10,11,12,13,14,15,16,17"
bashcommand_columns = "awk -F' ' '{print $0

"cut -d ' ' -f 0,3,5,6,8,9,10,11,12,13,14,15,16,17"

# load file and add worm id

In [105]:
# define columns to keep
col_ind_keep = [0,3,5,6,8,9,10,11,12,13,14,15,16,17]

# load column names legend
filename = 'legend_trinity_worm.pickle'
v = pickle.load(open(os.path.join(dir_save, filename),'rb'))

p = os.path.join(dir_save, 'test_cat_csv.csv')
df = pd.read_csv(p, usecols=col_ind_keep, delim_whitespace=True, 
                 header=None, 
                 names=v.loc[col_ind_keep,'name'].values)
# add worm id
df['worm_id'] = worm_id_array

In [108]:
# add ethanol condition
df['ethanol'] = np.tile(0,df.shape[0])

time   speed  bias  tap    loc_x    loc_y  morphwidth  midline  \
0      691.948     NaN   1.0  0.0  13.6952  28.7706      0.0850   0.6821   
1      692.034  0.0563   1.0  0.0  13.6888  28.7758      0.0797   0.6892   
2      692.118  0.0664   1.0  0.0  13.6877  28.7767      0.0845   0.6892   
3      692.185  0.0578   1.0  0.0  13.6814  28.7813      0.0819   0.6746   
4      692.277  0.0239   1.0  0.0  13.6814  28.7812      0.0805   0.6804   
...        ...     ...   ...  ...      ...      ...         ...      ...   
5911  1189.451  0.0129   0.0  0.0  18.0340  27.9478      0.1132   1.1996   
5912  1189.556  0.0036   0.0  0.0  18.0343  27.9491      0.1101   1.1796   
5913  1189.636  0.0252   0.0  0.0  18.0336  27.9477      0.1129   1.1976   
5914  1189.717     NaN   0.0  0.0  18.0344  27.9451      0.1106   1.1834   
5915  1189.798     NaN   0.0  0.0  18.0324  27.9437      0.1067   1.1558   

          area  angular  aspect  kink  curve    crab  worm_id  ethanol  
0     0.075816      NaN   0.202  73.4   29.7     NaN    12099        0  
1     0.071442      8.1   0.175  65.7   27.1  0.0039    12099        0  
2     0.072900     10.3   0.171  66.3   28.1  0.0027    12099        0  
3     0.071442      7.0   0.178  63.7   26.9  0.0028    12099        0  
4     0.070713      3.5   0.182  67.1   26.5  0.0035    12099        0  
...        ...      ...     ...   ...    ...     ...      ...      ...  
5911  0.172044      2.1   0.113  18.1   22.1  0.0080    18168        0  
5912  0.167670      1.7   0.115  19.6   21.5  0.0014    18168        0  
5913  0.173502      1.7   0.147  16.3   22.9  0.0124    18168        0  
5914  0.169857      NaN   0.151  15.9   22.7     NaN    18168        0  
5915  0.162567      NaN   0.183  17.7   22.3     NaN    18168        0  

[5916 rows x 16 columns]

In [ ]:
# creat group names

# Putting all together

In [128]:
# get file info file --
filename = 'fileinfo_trinity_N2400mM.pickle'
fileinfoDB = pickle.load(open(os.path.join(dir_save, filename),'rb'))


In [ ]:
df_target = fileDB_target.sample(10, random_state=318).copy()

In [154]:
# concat --
# define output name
output_filname = 'test_cat_csv.csv'

# create bash command components
bashcommand_prefix = 'cat'
bashcommand_sourcepaths = ''
bashcommand_space = ' '
bashcommand_link = '>'
bashcommand_outputpath = '"'+os.path.join(dir_save, output_filname)+'"'
# link file paths
for p in df_target['path'].values:
    p = '"'+p+'"'
    bashcommand_sourcepaths += p
    bashcommand_sourcepaths += ' '
# create bash command
bashcommand = bashcommand_prefix + bashcommand_space + \
            bashcommand_sourcepaths + \
            bashcommand_link + bashcommand_space + \
            bashcommand_outputpath
# call bash command
os.system(bashcommand)

0

In [160]:
# create wormid array and ethanol array --
# get info
worm_id_list = df_target['worm_id'].values
ethanol_list = df_target['ethanol'].values
rows_list = df_target['row_number'].values
row_total = df_target['row_number'].sum()
# instantiate arrays
worm_id_array = np.empty([row_total],dtype='int')
ethanol_array = np.empty([row_total],dtype='int')

# create arrays
i_row_previous = 0
for worm_id, ethanol, rows in zip(worm_id_list, ethanol_list, rows_list):
    # get start row position
    i_start_row = i_row_previous
    # get end row position
    i_end_row = i_row_previous+rows-1
    # get worm id array x number of rows
    worm_id_array[i_start_row:i_end_row+1] = np.tile(worm_id, rows)
    # get ethanol array x number of rows
    ethanol_array[i_start_row:i_end_row+1] = np.tile(ethanol, rows)
    # create next starting position
    i_row_previous = rows+i_row_previous

In [161]:
worm_id_array

array([11658, 11658, 11658, ...,  2742,  2742,  2742])

In [175]:
# save to space deliminated without header --
path_id_array = os.path.join(dir_save, 'temp_wormid_eth.csv')
# create data frame
df = pd.DataFrame({'worm_id':worm_id_array, 
                   'ethanol':ethanol_array})
# save csv
df.to_csv(path_id_array, sep=' ',index=False, header=False)



In [182]:
# concat identifier arrays with main array

bashcommand = "paste '" + \
            os.path.join(dir_save, 'temp_wormid_eth.csv') + \
            "' '" + \
            os.path.join(dir_save, 'test_cat_csv.csv') + \
            "' > '" + \
            os.path.join(dir_save, 'test_horcat.csv') + \
            "'"
bashcommand

"paste '/Users/connylin/Dropbox/CA/ED _20200119 Brain Station Data Science Diploma/Capstone/data/temp_wormid_eth.csv' '/Users/connylin/Dropbox/CA/ED _20200119 Brain Station Data Science Diploma/Capstone/data/test_cat_csv.csv' > '/Users/connylin/Dropbox/CA/ED _20200119 Brain Station Data Science Diploma/Capstone/data/test_horcat.csv'"

In [183]:
os.system(bashcommand)

0

In [190]:
col_ind_keep = [0,1,2,5,7,8,10,11,12,13,14,15,16,17,18,19]

In [194]:
# try loading 
# load column names legend
p = os.path.join(dir_save, 'legend_trinity_worm.pickle')
cnames = pickle.load(open(p,'rb'))
cnames = cnames.loc[[0,3,5,6,8,9,10,11,12,13,14,15,16,17],'name'].values
cnames = np.hstack((np.array(['worm_id','ethanol']),cnames)) 

In [195]:
p = os.path.join(dir_save, 'test_horcat.csv')
df = pd.read_csv(p, delim_whitespace=True, header=None,
                 usecols=[0,1,2,5,7,8,10,11,12,13,14,15,16,17,18,19], 
                 names=cnames)

worm_id  ethanol     time   speed  bias  tap    loc_x    loc_y  \
0       11658        1  739.204     NaN   1.0  0.0  13.3701  43.8153   
1       11658        1  739.257  0.2944   1.0  0.0  13.3588  43.8030   
2       11658        1  739.330  0.2409   1.0  0.0  13.3454  43.7881   
3       11658        1  739.394  0.1596   1.0  0.0  13.3375  43.7775   
4       11658        1  739.457  0.1637   1.0  0.0  13.3333  43.7718   
...       ...      ...      ...     ...   ...  ...      ...      ...   
8344     2742        0  309.371  0.1597   1.0  0.0  22.4471  50.4527   
8345     2742        0  309.481  0.1760   0.0  0.0  22.4439  50.4740   
8346     2742        0  309.604  0.1727   0.0  0.0  22.4456  50.4934   
8347     2742        0  309.735     NaN   0.0  0.0  22.4454  50.5180   
8348     2742        0  309.855     NaN   0.0  0.0  22.4463  50.5477   

      morphwidth  midline      area  angular  aspect  kink  curve    crab  
0         0.0903   1.2556  0.155277      NaN   0.104  12.8   16.3     NaN  
1         0.0899   1.2507  0.156006      2.2   0.124  17.3   16.8  0.0121  
2         0.0902   1.2357  0.153819      2.1   0.120  18.4   16.4  0.0017  
3         0.0902   1.2504  0.153819      2.3   0.111  20.0   16.2  0.0106  
4         0.0903   1.2591  0.155277      2.8   0.126  23.2   15.7  0.0100  
...          ...      ...       ...      ...     ...   ...    ...     ...  
8344      0.1008   0.7826  0.096228      0.4   0.392  40.8   28.1  0.0415  
8345      0.0970   0.8281  0.099144      8.8   0.339  33.9   28.9  0.0343  
8346      0.0906   0.8871  0.100602     11.0   0.322  36.4   23.1  0.0162  
8347      0.0917   0.8637  0.098415      NaN   0.316  53.6   24.8     NaN  
8348      0.1016   0.8854  0.107163      NaN   0.341  34.1   30.9     NaN  

[8349 rows x 16 columns]

In [11]:
def load_trinity_df(file_path, dir_save):
    ci = [0,3,5,6,8,9,10,11,12,13,14,15,16,17]
    # load column names legend
    p = os.path.join(dir_save, 'legend_trinity_worm.pickle')
    cnames = pickle.load(open(p,'rb'))
    cnames = cnames.loc[ci,'name'].values
    cnames = np.hstack((np.array(['worm_id','ethanol']),cnames)) 
    
    p = os.path.join(dir_save, file_path)
    df = pd.read_csv(p, delim_whitespace=True, header=None,
                 usecols=ci, 
                 names=cnames)

# issues with concating paths. double check if all paths points to right trinity sizes


## maybe there are zero size

In [4]:
# load pickle
filename = 'fileinfo_trinity_N2400mM.pickle'
fileDB_target = pickle.load(open(os.path.join(dir_save, filename),'rb'))

In [7]:
any(fileDB_target['row_number']<=1)


True

In [8]:
sum(fileDB_target['row_number']<=1)

6

In [17]:
i = fileDB_target['row_number']<=1
fileDB_target.iloc[i.values,:]

Unnamed: 0 dbname                                expname groupname  \
133476       133476    MWT             20120309X_CL_100s30x60s10s        N2   
264459       264459    MWT             20120609M_BM_100s30x60s10s        N2   
555117       555117    MWT  20131020A_SJ_100s30x60s10s_Neuroligin        N2   
580018       580018    MWT  20131020B_SJ_100s30x60s10s_Neuroligin        N2   
582035       582035    MWT  20131020B_SJ_100s30x60s10s_Neuroligin  N2_400mM   
1159802     1159802    MWT    20140809C_SJ_100s30x60s10s_glr1_60s        N2   

                 mwtname                                           filename  \
133476   20120309_182530  N2_8x3_l92h20C_100s30x60s60s_C0309ac.trinity.3...   
264459   20120609_164131  N2_5x3_t90h20C_100s30x60s60s_C1029bg.trinity.1...   
555117   20131020_114017  N2_5x3_t96h20C_100s30x10s10s_C0109ec.trinity.0...   
580018   20131020_123616  N2_5x3_t96h20C_100s30x10s10s_C0109ad.trinity.1...   
582035   20131020_131753  N2_5x3_t96h20C_100s30x10s10s_C0109ba.trinity.2...   
1159802  20140809_190234  N2_5x3_t96h20C_100s30x10s10s_C0109ad.trinity.2...   

                                                      path  \
133476   /Volumes/COBOLT/MWT/20120309X_CL_100s30x60s10s...   
264459   /Volumes/COBOLT/MWT/20120609M_BM_100s30x60s10s...   
555117   /Volumes/COBOLT/MWT/20131020A_SJ_100s30x60s10s...   
580018   /Volumes/COBOLT/MWT/20131020B_SJ_100s30x60s10s...   
582035   /Volumes/COBOLT/MWT/20131020B_SJ_100s30x60s10s...   
1159802  /Volumes/COBOLT/MWT/20140809C_SJ_100s30x60s10s...   

                                                     fname   ext  \
133476   N2_8x3_l92h20C_100s30x60s60s_C0309ac.trinity.3...  .dat   
264459   N2_5x3_t90h20C_100s30x60s60s_C1029bg.trinity.1...  .dat   
555117   N2_5x3_t96h20C_100s30x10s10s_C0109ec.trinity.0...  .dat   
580018   N2_5x3_t96h20C_100s30x10s10s_C0109ad.trinity.1...  .dat   
582035   N2_5x3_t96h20C_100s30x10s10s_C0109ba.trinity.2...  .dat   
1159802  N2_5x3_t96h20C_100s30x10s10s_C0109ad.trinity.2...  .dat   

                  analysis_tag  worm_id  row_number  ethanol  
133476   ['.32438' '.trinity']    32438           1        0  
264459   ['.10447' '.trinity']    10447           1        0  
555117   ['.08917' '.trinity']     8917           1        0  
580018   ['.11188' '.trinity']    11188           1        0  
582035   ['.28676' '.trinity']    28676           1        1  
1159802  ['.24485' '.trinity']    24485           1        0

In [19]:
file_path = fileDB_target.loc[555117,'path']

pd.read_csv(file_path, delim_whitespace=True, header=None)

0    1    2   3    4    5    6    7        8        9      10      11  \
0  294.912  1.0  1.0 NaN  0.0  1.0  0.0  0.0  44.2236  13.4856  0.109  0.2654   

        12  13     14    15    16  17  
0  0.03645 NaN  0.566  76.9  24.8 NaN

In [6]:
# look at column size for each file
fileDB_target

Unnamed: 0 dbname                            expname groupname  \
39135         39135    MWT   20110906B_CL_100s30x60s10s_pilot        N2   
39136         39136    MWT   20110906B_CL_100s30x60s10s_pilot        N2   
39137         39137    MWT   20110906B_CL_100s30x60s10s_pilot        N2   
39138         39138    MWT   20110906B_CL_100s30x60s10s_pilot        N2   
39139         39139    MWT   20110906B_CL_100s30x60s10s_pilot        N2   
...             ...    ...                                ...       ...   
1307197     1307197    MWT  20170605X_CR_100s30x10s10s_TM1630  N2_400mM   
1307198     1307198    MWT  20170605X_CR_100s30x10s10s_TM1630  N2_400mM   
1307199     1307199    MWT  20170605X_CR_100s30x10s10s_TM1630  N2_400mM   
1307200     1307200    MWT  20170605X_CR_100s30x10s10s_TM1630  N2_400mM   
1307201     1307201    MWT  20170605X_CR_100s30x10s10s_TM1630  N2_400mM   

                 mwtname                                           filename  \
39135    20110906_113945  N2_5x3_f92h20C_100s30x60s10s_B0902bb.trinity.0...   
39136    20110906_113945  N2_5x3_f92h20C_100s30x60s10s_B0902bb.trinity.0...   
39137    20110906_113945  N2_5x3_f92h20C_100s30x60s10s_B0902bb.trinity.0...   
39138    20110906_113945  N2_5x3_f92h20C_100s30x60s10s_B0902bb.trinity.0...   
39139    20110906_113945  N2_5x3_f92h20C_100s30x60s10s_B0902bb.trinity.0...   
...                  ...                                                ...   
1307197  20170605_153818  C8_tm1630_30x0_t96h20C_100s30x10s10s_C0605_CMR...   
1307198  20170605_153818  C8_tm1630_30x0_t96h20C_100s30x10s10s_C0605_CMR...   
1307199  20170605_153818  C8_tm1630_30x0_t96h20C_100s30x10s10s_C0605_CMR...   
1307200  20170605_153818  C8_tm1630_30x0_t96h20C_100s30x10s10s_C0605_CMR...   
1307201  20170605_153818  C8_tm1630_30x0_t96h20C_100s30x10s10s_C0605_CMR...   

                                                      path  \
39135    /Volumes/COBOLT/MWT/20110906B_CL_100s30x60s10s...   
39136    /Volumes/COBOLT/MWT/20110906B_CL_100s30x60s10s...   
39137    /Volumes/COBOLT/MWT/20110906B_CL_100s30x60s10s...   
39138    /Volumes/COBOLT/MWT/20110906B_CL_100s30x60s10s...   
39139    /Volumes/COBOLT/MWT/20110906B_CL_100s30x60s10s...   
...                                                    ...   
1307197  /Volumes/COBOLT/MWT/20170605X_CR_100s30x10s10s...   
1307198  /Volumes/COBOLT/MWT/20170605X_CR_100s30x10s10s...   
1307199  /Volumes/COBOLT/MWT/20170605X_CR_100s30x10s10s...   
1307200  /Volumes/COBOLT/MWT/20170605X_CR_100s30x10s10s...   
1307201  /Volumes/COBOLT/MWT/20170605X_CR_100s30x10s10s...   

                                                     fname   ext  \
39135    N2_5x3_f92h20C_100s30x60s10s_B0902bb.trinity.0...  .dat   
39136    N2_5x3_f92h20C_100s30x60s10s_B0902bb.trinity.0...  .dat   
39137    N2_5x3_f92h20C_100s30x60s10s_B0902bb.trinity.0...  .dat   
39138    N2_5x3_f92h20C_100s30x60s10s_B0902bb.trinity.0...  .dat   
39139    N2_5x3_f92h20C_100s30x60s10s_B0902bb.trinity.0...  .dat   
...                                                    ...   ...   
1307197  C8_tm1630_30x0_t96h20C_100s30x10s10s_C0605_CMR...  .dat   
1307198  C8_tm1630_30x0_t96h20C_100s30x10s10s_C0605_CMR...  .dat   
1307199  C8_tm1630_30x0_t96h20C_100s30x10s10s_C0605_CMR...  .dat   
1307200  C8_tm1630_30x0_t96h20C_100s30x10s10s_C0605_CMR...  .dat   
1307201  C8_tm1630_30x0_t96h20C_100s30x10s10s_C0605_CMR...  .dat   

                  analysis_tag  worm_id  row_number  ethanol  
39135    ['.00001' '.trinity']        1         527        0  
39136    ['.00002' '.trinity']        2         915        0  
39137    ['.04098' '.trinity']     4098        9648        0  
39138    ['.00004' '.trinity']        4        1611        0  
39139    ['.00005' '.trinity']        5         707        0  
...                        ...      ...         ...      ...  
1307197  ['.02552' '.trinity']     2552        1186        1  
1307198  ['.00505' '.trinity']      505         477        1  
1307199  ['.02041' '.trinity']

## go through each file

In [18]:
ci = [0,3,5,6,8,9,10,11,12,13,14,15,16,17]
# load column names legend
p = os.path.join(dir_save, 'legend_trinity_worm.pickle')
cnames = pickle.load(open(p,'rb'))
cnames = cnames['name'].values
# cnames = cnames.loc[ci,'name'].values
# cnames = np.hstack((np.array(['worm_id','ethanol']),cnames)) 
# cnames

In [20]:
fileDB_target

Unnamed: 0 dbname                            expname groupname  \
39135         39135    MWT   20110906B_CL_100s30x60s10s_pilot        N2   
39136         39136    MWT   20110906B_CL_100s30x60s10s_pilot        N2   
39137         39137    MWT   20110906B_CL_100s30x60s10s_pilot        N2   
39138         39138    MWT   20110906B_CL_100s30x60s10s_pilot        N2   
39139         39139    MWT   20110906B_CL_100s30x60s10s_pilot        N2   
...             ...    ...                                ...       ...   
1307197     1307197    MWT  20170605X_CR_100s30x10s10s_TM1630  N2_400mM   
1307198     1307198    MWT  20170605X_CR_100s30x10s10s_TM1630  N2_400mM   
1307199     1307199    MWT  20170605X_CR_100s30x10s10s_TM1630  N2_400mM   
1307200     1307200    MWT  20170605X_CR_100s30x10s10s_TM1630  N2_400mM   
1307201     1307201    MWT  20170605X_CR_100s30x10s10s_TM1630  N2_400mM   

                 mwtname                                           filename  \
39135    20110906_113945  N2_5x3_f92h20C_100s30x60s10s_B0902bb.trinity.0...   
39136    20110906_113945  N2_5x3_f92h20C_100s30x60s10s_B0902bb.trinity.0...   
39137    20110906_113945  N2_5x3_f92h20C_100s30x60s10s_B0902bb.trinity.0...   
39138    20110906_113945  N2_5x3_f92h20C_100s30x60s10s_B0902bb.trinity.0...   
39139    20110906_113945  N2_5x3_f92h20C_100s30x60s10s_B0902bb.trinity.0...   
...                  ...                                                ...   
1307197  20170605_153818  C8_tm1630_30x0_t96h20C_100s30x10s10s_C0605_CMR...   
1307198  20170605_153818  C8_tm1630_30x0_t96h20C_100s30x10s10s_C0605_CMR...   
1307199  20170605_153818  C8_tm1630_30x0_t96h20C_100s30x10s10s_C0605_CMR...   
1307200  20170605_153818  C8_tm1630_30x0_t96h20C_100s30x10s10s_C0605_CMR...   
1307201  20170605_153818  C8_tm1630_30x0_t96h20C_100s30x10s10s_C0605_CMR...   

                                                      path  \
39135    /Volumes/COBOLT/MWT/20110906B_CL_100s30x60s10s...   
39136    /Volumes/COBOLT/MWT/20110906B_CL_100s30x60s10s...   
39137    /Volumes/COBOLT/MWT/20110906B_CL_100s30x60s10s...   
39138    /Volumes/COBOLT/MWT/20110906B_CL_100s30x60s10s...   
39139    /Volumes/COBOLT/MWT/20110906B_CL_100s30x60s10s...   
...                                                    ...   
1307197  /Volumes/COBOLT/MWT/20170605X_CR_100s30x10s10s...   
1307198  /Volumes/COBOLT/MWT/20170605X_CR_100s30x10s10s...   
1307199  /Volumes/COBOLT/MWT/20170605X_CR_100s30x10s10s...   
1307200  /Volumes/COBOLT/MWT/20170605X_CR_100s30x10s10s...   
1307201  /Volumes/COBOLT/MWT/20170605X_CR_100s30x10s10s...   

                                                     fname   ext  \
39135    N2_5x3_f92h20C_100s30x60s10s_B0902bb.trinity.0...  .dat   
39136    N2_5x3_f92h20C_100s30x60s10s_B0902bb.trinity.0...  .dat   
39137    N2_5x3_f92h20C_100s30x60s10s_B0902bb.trinity.0...  .dat   
39138    N2_5x3_f92h20C_100s30x60s10s_B0902bb.trinity.0...  .dat   
39139    N2_5x3_f92h20C_100s30x60s10s_B0902bb.trinity.0...  .dat   
...                                                    ...   ...   
1307197  C8_tm1630_30x0_t96h20C_100s30x10s10s_C0605_CMR...  .dat   
1307198  C8_tm1630_30x0_t96h20C_100s30x10s10s_C0605_CMR...  .dat   
1307199  C8_tm1630_30x0_t96h20C_100s30x10s10s_C0605_CMR...  .dat   
1307200  C8_tm1630_30x0_t96h20C_100s30x10s10s_C0605_CMR...  .dat   
1307201  C8_tm1630_30x0_t96h20C_100s30x10s10s_C0605_CMR...  .dat   

                  analysis_tag  worm_id  row_number  ethanol  
39135    ['.00001' '.trinity']        1         527        0  
39136    ['.00002' '.trinity']        2         915        0  
39137    ['.04098' '.trinity']     4098        9648        0  
39138    ['.00004' '.trinity']        4        1611        0  
39139    ['.00005' '.trinity']        5         707        0  
...                        ...      ...         ...      ...  
1307197  ['.02552' '.trinity']     2552        1186        1  
1307198  ['.00505' '.trinity']      505         477        1  
1307199  ['.02041' '.trinity']

In [19]:

for p in fileDB_target.index.values:
    print(p)
    df = pd.read_csv(p, delim_whitespace=True, header=None, names=cnames)
    display(df)
    break

/Volumes/COBOLT/MWT/20110906B_CL_100s30x60s10s_pilot/N2/20110906_113945/N2_5x3_f92h20C_100s30x60s10s_B0902bb.trinity.00001.dat


time  number  goodnumber   speed  speed_std  bias  tap  puff   loc_x  \
0     3.352     1.0         1.0     NaN        0.0   1.0  0.0   0.0  2.1315   
1     3.442     1.0         1.0  0.3676        0.0   1.0  0.0   0.0  2.1230   
2     3.513     1.0         1.0  0.3751        0.0   1.0  0.0   0.0  2.1152   
3     3.593     1.0         1.0  0.3542        0.0   1.0  0.0   0.0  2.1079   
4     3.677     1.0         1.0  0.3700        0.0   1.0  0.0   0.0  2.1002   
..      ...     ...         ...     ...        ...   ...  ...   ...     ...   
522  36.240     1.0         1.0  0.3513        0.0   1.0  0.0   0.0  3.2782   
523  36.279     1.0         1.0  0.4126        0.0   0.0  0.0   0.0  3.2875   
524  36.319     1.0         1.0     NaN        0.0   0.0  0.0   0.0  3.2956   
525  36.361     1.0         1.0     NaN        0.0   0.0  0.0   0.0  3.3047   
526  36.396     1.0         1.0     NaN        0.0   0.0  0.0   0.0  3.3149   

       loc_y  morphwidth  midline      area  angular  aspect  kink  curve  \
0    30.6374      0.1239   1.2712  0.172044      NaN   0.190  35.1   31.4   
1    30.6657      0.1221   1.2892  0.171315      3.8   0.194  31.3   32.4   
2    30.6935      0.1263   1.2925  0.175689      3.9   0.200  29.2   28.2   
3    30.7200      0.1227   1.2818  0.173502      3.4   0.175  30.4   28.4   
4    30.7497      0.1203   1.2765  0.169857      2.4   0.173  29.3   26.1   
..       ...         ...      ...       ...      ...     ...   ...    ...   
522  41.2798      0.1045   1.3244  0.174960      4.0   0.136  33.3   29.6   
523  41.2952      0.1030   1.3125  0.174231      4.8   0.134  35.9   30.4   
524  41.3096      0.1022   1.2769  0.172044      NaN   0.137  34.9   27.9   
525  41.3235      0.1030   1.2923  0.173502      NaN   0.192  40.2   29.2   
526  41.3411      0.1045   1.3676  0.176418      NaN   0.199  39.9   28.9   

       crab  
0       NaN  
1    0.0038  
2    0.0101  
3    0.0140  
4    0.0256  
..      ...  
522  0.0153  
523  0.0216  
524     NaN  
525     NaN  
526     NaN  

[527 rows x 18 columns]

# Explore .mat trinitySummary.mat

In [22]:
# load file summary
file_summary = pickle.load(open(os.path.join(dir_save, 'file_summary_mwt.pickle'),'rb'))

In [30]:
i = file_summary['filepath']['.mat'].isna().values
pMWT = file_summary.index[~i].values

In [37]:
from scipy.io import loadmat

for p in pMWT:
    pTrinity = os.path.join(p,'trinitySummary.mat')
    p_exist = os.path.exists(pTrinity)
    print(pTrinity, p_exist)
    if p_exist:
        annots = loadmat(pTrinity)
        print(annots)


    break

/Volumes/COBOLT/MWT DISCARD/20120229B_DC_100s30x10s10s_training/N2/20120229_171242/trinitySummary.mat True


NotImplementedError: Please use HDF reader for matlab v7.3 files

# complie trinity.id.dat to one file

In [56]:
import subprocess


completeng #400 files


In [89]:
# get links to each plate --
# load pickle
plateDB = pickle.load(open(os.path.join(dir_save, 'file_summary_mwt.pickle'),'rb'))
# get paths with trinity.id.dat
pMWT = plateDB.index[~plateDB[('filepath','trinity.id.dat')].isna()].values

col_ind_keep = [0,3,5,6,8,9,10,11,12,13,14,15,16,17]
# load column names legend
p = os.path.join(dir_save, 'legend_trinity_worm.pickle')
cnames = pickle.load(open(p,'rb'))
cnames = cnames['name'].values
cnames_keep = cnames[col_ind_keep].copy()


In [101]:
# go in each plate
# for p in pMWT:
for plate_count, pPlate in enumerate(pMWT,1):
    print(f'processing {plate_count}/{pMWT.shape[0]} plate')
    # find all trinity files
    ptri = glob.glob(pPlate+"/*.trinity.*.dat")
    # print number of files
    print(f'\t{len(ptri)} trinity files in this plate')


    # get worm id from filep path
    df = pd.DataFrame({'paths':ptri})
    worm_number = df['paths'].str.extract(r'(?<=trinity[.])(\d{1,})')
    worm_number = worm_number.astype('int32').values

    # get number of rows per doc
    # declare empty array to store the row numbers
    rows_array = np.empty((len(ptri)),dtype='int')
    # loop through each files 
    for i, p in enumerate(ptri, 0):
        # report process every 100 files
        if (i%100==0):
            print(f'\tprocessing #{i} files', end='\r')
        # create custom bash command
        bashCommand = "cat "+p+" | wc -l"
        # call bash
        output = subprocess.check_output(bashCommand, shell=True)
        # take out digits from the byte output
        rows_array[i] = int(output)


    # create worm id array
    # get info
    row_total = rows_array.sum()
    # instantiate arrays
    row_repeats_array = np.empty([row_total], dtype='int')
    # create arrays
    i_row_previous = 0
    for worm_id, rows in zip(worm_number, rows_array):
        # get start row position
        i_start_row = i_row_previous
        # get end row position
        i_end_row = i_row_previous+rows-1
        # get worm id array x number of rows
        row_repeats_array[i_start_row:i_end_row+1] = np.tile(worm_id, rows)
        # create next starting position
        i_row_previous = rows+i_row_previous

    # concat paths
    bashcommand_sourcepaths = ''
    for p in ptri:
        p = "'"+p+"'"
        bashcommand_sourcepaths += p
        bashcommand_sourcepaths += ' '
    # create bash command components
    psave_trinity = os.path.join(pPlate, 'temp_trinity.dat')
    bashcommand = 'cat ' + bashcommand_sourcepaths + '> ' + "'"+ psave_trinity + "'"
    # call bash command
    os.system(bashcommand)

    # load tirnity concated
    df = pd.read_csv(psave_trinity, delim_whitespace=True, header=None, usecols=col_ind_keep, names=cnames_keep)
    # add wormid field
    df['wormid'] = row_repeats_array

    # into pickle
    psave_pickle = os.path.join(pPlate, 'trinity_all_worms.pickle')
    pickle.dump(df, open(psave_pickle,'wb'))
    if os.path.exists(psave_pickle):
        size_trinity = os.path.getsize(psave_pickle)
        print(f'\tsize of pickle file: {size_trinity/1000**2:.0f} MB')
        if os.path.exists(psave_trinity):
            print('\tclean up trinity temp file')
            os.remove(psave_trinity)
    else:
        print('\tpickle file did not save')

408
processing #400 files

size of pickle file: 99 MB
estimate total size: 86.28 GB
clean up trinity temp file


In [97]:
pMWT.shape[0]

871

# Problem with Bash cat limit - process by 1000 per batch

In [ ]:
# ------